File: src/aizk/graph/ids.py
import uuid

from pydantic import UUID5

from .naming import normalize_name

_NAMESPACE = uuid.UUID("a12c0de0-0000-5000-8000-a12c00000000")
_DELIMITER = "\x1f"


def normalize(value: str) -> str:
    """Fold a field to its canonical form before hashing, lowercased with collapsed
    whitespace."""
    return " ".join(value.split()).casefold()


def entity_id(name: str, type: str) -> UUID5:
    """Deterministic uuid5 for an entity from its normalized name and type."""
    fields = _DELIMITER.join((normalize(type), normalize_name(name)))
    return uuid.uuid5(_NAMESPACE, fields)


def fact_id(
    subject_id: UUID5,
    predicate: str,
    object_id: UUID5 | None,
    statement: str,
) -> UUID5:
    """Derive fact identity from resolved endpoints, predicate, and statement."""
    parts = (
        str(subject_id),
        normalize(predicate),
        str(object_id or ""),
        normalize(statement),
    )
    return uuid.uuid5(_NAMESPACE, _DELIMITER.join(parts))

File: archive/ts/src/lib/server/graph/ids.ts
// Deterministic content ids ported exactly from graph/ids.py: uuid5 over the same private
// namespace and unit-separator-joined normalized fields, so the same entity or fact minted
// from either server lands on the byte-identical content row.
import { v5 as uuid5 } from 'uuid';

import { casefold, collapseWhitespace } from '../extract/grounding';
import { normalizeName } from './naming';

const NAMESPACE = 'a12c0de0-0000-5000-8000-a12c00000000';
const DELIMITER = '\x1f';

// Fold a field to its canonical form before hashing, lowercased with collapsed whitespace.
const normalize = (value: string): string => casefold(collapseWhitespace(value));

export const entityId = (name: string, type: string): string =>
	uuid5([normalize(type), normalizeName(name)].join(DELIMITER), NAMESPACE);

export const factId = (
	subject: string,
	predicate: string,
	object: string,
	statement: string
): string =>
	uuid5(
		[normalize(subject), normalize(predicate), normalize(object), normalize(statement)].join(
			DELIMITER
		),
		NAMESPACE
	);
